### PR TITLE
Table Resolver

### PR DESCRIPTION
**Pull request by nablaone**
This PR introduces `Table Resolver`. Table resolver answers the question "What will Quesma do for given index (or index pattern)?" 

It has three parts:
- `TableResolver` interface and implementation
- set of rules
- UI 

Limitations: 
It doesn't perform any actions based on decision. Rules can be incomplete or simply wrong. It will be finished in the next PR.

### CODE REVIEW

File: quesma/table_resolver/table_resover.go [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+	"time"
+)
+
+type tableResolver interface {
+	resolve(indexPattern string) *Decision
+}
+
+type pattern struct {
+	// source
+	pattern string
```
Comment (pdelewski) [inline quesma/table_resolver/table_resover.go:25]: names of fields are repetitive and might be confusing

File: quesma/table_resolver/table_resover.go [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+
+	// parsed data
+	isPattern bool
+	patterns  []string
+}
+
+type namedResolver struct {
```
Comment (pdelewski) [inline quesma/table_resolver/table_resover.go:32]: maybe `basicResolver` or `simpleResolver`?

File: quesma/table_resolver/table_resover.go [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+	name     string
+	resolver func(pattern pattern) *Decision
+}
+
+type composedResolver struct {
```
Comment (pdelewski) [inline quesma/table_resolver/table_resover.go:37]: maybe `compoundResolver` ?

File: quesma/table_resolver/table_resover.go [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+		elasticIndexResolver: elasticResolver,
+		pipelineResolvers:    make(map[string]*pipelineResolver),
+	}
+
+	ingestResolver := &pipelineResolver{
+		pipelineName: IngestPipeline,
+		cfg:          ingestConf,
+		resolver: &composedResolver{
+			decisionLadder: []namedResolver{
+				{"patternIsNotAllowed", patternIsNotAllowed},
```
Comment (pdelewski) [inline quesma/table_resolver/table_resover.go:277]: Where these callbacks are defined like `patternIsNotAllowed`?

Comment (pdelewski) [reply quesma/table_resolver/table_resover.go:277]: ok, I see, rules.

File: quesma/table_resolver/rules.go [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+func makeIsDisabledInConfig(cfg map[string]config.IndexConfiguration, pipeline string) func(input parsedPattern) *Decision {
+
+	return func(input parsedPattern) *Decision {
+
+		if input.isPattern {
+
+			// pass to the next resolver
+
+			return nil
+		} else {
```
Comment (pdelewski) [inline quesma/table_resolver/rules.go:36]: Maybe, we could just do
```
if !input.isPattern
```

File: quesma/quesma/search.go [hunk 6]
```diff
@@ -265,6 +271,10 @@ func (q *QueryRunner) executePlan(ctx context.Context, plan *model.ExecutionPlan
 }
 
 func (q *QueryRunner) handleSearchCommon(ctx context.Context, indexPattern string, body types.JSON, optAsync *AsyncQuery, queryLanguage QueryLanguage) ([]byte, error) {
+
+	decision := q.tableResolver.Resolve(table_resolver.QueryPipeline, indexPattern)
+	table_resolver.TODO("handleSearchCommon", decision)
```
Comment (pdelewski) [inline quesma/quesma/search.go:276]: Do you already thought about dispatch implementation? (functionality of current `TODOs`)

File: quesma/table_resolver/model.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+func (*ConnectorDecisionElastic) Message() string {
+	return "Pass to Elasticsearch."
+}
+
+type ConnectorDecisionClickhouse struct {
+	// TODO  instance of clickhouse connector
+
+	ClickhouseTableName string
+	Indexes             []string
```
Comment (pdelewski) [inline quesma/table_resolver/model.go:66]: Are these indexes elastic ones?

Comment (nablaone) [reply quesma/table_resolver/model.go:66]: These are matched indexes. I'll change the name.

File: quesma/quesma/search.go (hunk 6 above)
Comment (nablaone) [reply quesma/quesma/search.go:276]: `TODO` is a marker function. It will be removed.

Comment (pdelewski) [reply quesma/quesma/search.go:276]: I know it's marker but you need to do something based on decision and I'm asking about this implementation

Comment (nablaone) [reply quesma/quesma/search.go:276]: I had a vision, but I can't tell you how it will be implemented.

File: quesma/table_resolver/model.go (hunk 7 above)
Comment (pdelewski) [reply quesma/table_resolver/model.go:66]: So maybe cleaner would be to change name to `ClickhouseIndexes` or `ClickhouseTables`?

**Review by pdelewski — APPROVED** [review]

**Pull request merged**